docs/training_resources/python/data-classes.md

```diff
@@ -0,0 +1,151 @@
+
+Someone who wanted to use these functions could then do so as follows:
+
+```py
+start_date = date(2023, 1, 1)
+end_date = date(2024, 1, 1)
+my_publication_dates = PublicationDates(start_date, end_date)
+
+my_first_result = my_first_function(my_publication_dates)
+my_second_result = my_second_function(my_publication_dates)
+```
+
+Although the benefits may be limited for this contrived example, we can see that:
+
+1. The intent is clearer when we read our function - rather than just having independent parameters called `start_date` and `end_date` we immediately know that these variables are a pair, representing the publication start and end dates, since they are contained together within the data class.
+2. Since `my_publication_dates` is passed to both functions, we now have a single argument for each function, so the code is easier to read and there is less room for error (e.g. it's not possible to mix up our `start_date` and `end_date` . This doesn't make a huge difference in this example with two functions and two parameters, but imagine what happens as the pipeline gets bigger.
+
+## Validations using data classes
+
+A great feature of data classes is that after they are created, a special function - called `__postinit__` runs that you can use to do things like run validations or checks, or to derive additional attributes. 
+
+For example:
+
+```py
+import dataclasses
+from datetime import date
+
+@dataclasses.dataclass
+class PublicationDates:
+  start_date: date
+  end_date: date
+
+  def __postinit__(self):
+    # Raise an error if start date less than end date
+    assert self.start_date < self.end_date, "Start date was not less than end date!"
+
+good_publication_dates = PublicationDates(start_date=date(2022, 1, 1), end_date=date(2023, 1, 1)) # This will be fine
+bad_publication_dates = PublicationDates(start_date=date(2023, 1, 1), end_date=date(2022, 1, 1)) # This will raise an AssertionError
+```
+
+So now, in addition to the benefits we had previously, we can also have checks and validations running at the time that we define our parameters  - rather than individually within each function, or via creating another function to do these checks that we need to remember to call - the logic is built into how the parameters are defined! This also makes it clearer to someone who wants to use these parameters that there are constraints on the values.
+
+You don't have to define a `__postinit__` method but it can be useful in certain situations!
+
+## Further reading
+
+- [Data classes guide on Data Quest](https://www.dataquest.io/blog/how-to-use-python-data-classes/)
+- For NHSE Employees, see this [video walkthrough](https://nhsd-confluence.digital.nhs.uk/download/attachments/483240713/Classes%20in%20RAP%20projects%20demo-20230413_103354-Meeting%20Recording.mp4?api=v2&modificationDate=1681382735402&version=1) of how data classes could be used to improve RAP projects, with examples from the National Diabetes Audit RAP project.
```

Comment:
I can access this fine but I think maybe since it's on confluence only NHSD employees with a confluence login. Might need changing to "For NHSE with Confluence access". Not sure if that's the only barrier or if it's help within a certain area or something

Reply:
I'll just remove this, I can't access it either so I guess it's in a restricted team space. thanks!